### PR TITLE
feat(tree): query attribute and property

### PR DESCRIPTION
**Pull request by Sakchai-Refinitiv**
Ref: https://jira.refinitiv.com/browse/ELF-1500

### CODE REVIEW

**Comment by TremayneChrist** [comment]
@Sakchai-Refinitiv I am still reviewing this.

File: packages/elements/src/tree/elements/tree.ts [hunk 1]
```diff
@@ -254,10 +268,136 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
...
+          this.manager.includeItem(item);
+        }
+        else {
+          this.manager.excludeItem(item);
+        }
+
+        return result;
+      }).slice();
+
+      // Do not expand EMS if there is no filter applied
```
Comment (TremayneChrist) [inline packages/elements/src/tree/elements/tree.ts:306]: Copied comment from EMS

File: packages/elements/src/tree/elements/tree.ts [hunk 2]
```diff
@@ -254,10 +268,136 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
...
+      if (this.query) {
+        /**
+         * Include all descendants of item list to make them all
```
Comment (TremayneChrist) [inline packages/elements/src/tree/elements/tree.ts:309]: Comment needs rewording

File: packages/elements/src/tree/elements/tree.ts [hunk 3]
```diff
@@ -254,10 +268,136 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
...
+   * Utility method
+   * Add nested children of item list
+   * @param items List of items
+   * @param excludeItems List of exclude items
+   * @returns {void}
+   */
+  protected addNestedItemsToRender (items: readonly T[], excludeItems: readonly T[]): void {
+    items.forEach(item => {
+      // Skip excluding item
+      if (!excludeItems.includes(item)) {
```
Comment (TremayneChrist) [inline packages/elements/src/tree/elements/tree.ts:356]:
```suggestion
      if (!item.hidden && !excludeItems.includes(item)) {
```

Comment (TremayneChrist) [reply packages/elements/src/tree/elements/tree.ts:356]: You need this otherwise you could end up showing items that should be hidden by default

File: packages/elements/src/tree/elements/tree.ts [hunk 4]
```diff
@@ -254,10 +268,136 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
...
+        // Add item and nested children
+        if (this.composer.getItemPropertyValue(item, 'hidden') === true) {
```
Comment (TremayneChrist) [inline packages/elements/src/tree/elements/tree.ts:358]: Why do you check the hidden state here? Just include the item.

File: packages/elements/src/tree/helpers/filter.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { Tree, TreeDataItem } from '../index';
+import { TreeFilter } from './types';
+
+/**
+ * Default filter used by combo box
```
Comment (TremayneChrist) [inline packages/elements/src/tree/helpers/filter.ts:5]: Copied comments need fixing

File: packages/elements/src/tree/helpers/types.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import { CollectionComposer, DataItem } from '@refinitiv-ui/utils';
 
-export type TreeData<T extends TreeDataItem = TreeDataItem> = T[] | CollectionComposer<T> | null;
```
Comment (TremayneChrist) [inline packages/elements/src/tree/helpers/types.ts:3]: Why did you remove the default?

File: packages/elements/src/tree/__demo__/index.html [hunk 7]
```diff
@@ -147,6 +147,29 @@
...
+        const debounce = (callback, timeout = 500) => {
+          let timer = setTimeout(callback, timeout);
+          return (...args) => {
+            clearTimeout(timer);
+            timer = setTimeout(() => callback.apply(this, args), timeout);
+          };
+        };
+        const el = document.getElementById('queryTree');
+        const setQuery = debounce((e) => {
+          el.query = e.target.value;
```
Comment (TremayneChrist) [inline packages/elements/src/tree/__demo__/index.html:167]: This line is failing in the demo page on page load

**Review by TremayneChrist — CHANGES_REQUESTED** [review]
Hey @Sakchai-Refinitiv,

Please address these highlighted issues.

Can you also check the linting issues. I know they are not part of your changes, however, they can easily be fixed.

File: packages/elements/src/tree-select/__test__/tree-select.filter.test.js [hunk 8]
```diff
@@ -45,7 +45,41 @@ describe('tree-select/Filter', () => {
       expect(el.treeManager.visibleItems[1].label).to.equal('Republic of Macedonia');
     });
 
-    it('Text filter applied - multi level', async () => {
+    it('Text filter applied, expanded ancestors of matched items correctly - multi level', async () => {
+      const el = await fixture('<ef-tree-select></ef-tree-select>');
+      el.data = multiLevelData;
+      el.data = multiLevelData;
+      await openedUpdated(el);
```
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree-select/__test__/tree-select.filter.test.js:52]: In the `openUpdated(el)` wrap the `elementUpdated(el)` inside it.

File: packages/elements/src/tree-select/__test__/tree-select.filter.test.js [hunk 9]
```diff
@@ -45,7 +45,41 @@ describe('tree-select/Filter', () => {
...
+      await elementUpdated(el);
+      expect(el.treeManager.isItemExpanded(el.treeManager.visibleItems[0])).to.equal(true, 'Level 1-1 is expanded because matched some descendant item');
+      expect(el.treeManager.isItemExpanded(el.treeManager.visibleItems[1])).to.equal(true, 'Level 2-1 is expanded because matched some descendant item');
+    });
+
+    it('Text filter applied, collapsed children of matched items and included descendants correctly - multi level', async () => {
+      const el = await fixture('<ef-tree-select></ef-tree-select>');
+      el.data = multiLevelData;
+      el.data = multiLevelData;
+      await openedUpdated(el);
```
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree-select/__test__/tree-select.filter.test.js:65]: In the `openedUpdated` wrap the `elementUpdated` inside it.

File: packages/elements/src/tree/helpers/types.ts (hunk 6 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/helpers/types.ts:3]: Both of `TreeDataItem` are the same thing, no need to add the default, Am I understand correctly?

File: packages/elements/src/tree-select/__test__/tree-select.filter.test.js [hunk 10]
```diff
@@ -45,7 +45,41 @@ describe('tree-select/Filter', () => {
       expect(el.treeManager.visibleItems[1].label).to.equal('Republic of Macedonia');
     });
 
-    it('Text filter applied - multi level', async () => {
+    it('Text filter applied, expanded ancestors of matched items correctly - multi level', async () => {
+      const el = await fixture('<ef-tree-select></ef-tree-select>');
+      el.data = multiLevelData;
+      el.data = multiLevelData;
```
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree-select/__test__/tree-select.filter.test.js]: Removed

File: packages/elements/src/tree-select/__test__/tree-select.filter.test.js [hunk 11]
```diff
@@ -45,7 +45,41 @@ describe('tree-select/Filter', () => {
...
+      el.query = '-2';
+      await elementUpdated(el);
+      expect(el.treeManager.isItemExpanded(el.treeManager.visibleItems[0])).to.equal(true, 'Level 1-1 is expanded because matched some descendant item');
+      expect(el.treeManager.isItemExpanded(el.treeManager.visibleItems[1])).to.equal(true, 'Level 2-1 is expanded because matched some descendant item');
+    });
+
+    it('Text filter applied, collapsed children of matched items and included descendants correctly - multi level', async () => {
+      const el = await fixture('<ef-tree-select></ef-tree-select>');
+      el.data = multiLevelData;
+      el.data = multiLevelData;
```
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree-select/__test__/tree-select.filter.test.js]: Removed

File: packages/elements/src/tree-select/index.ts [hunk 12]
```diff
@@ -558,36 +558,37 @@ export class TreeSelect extends ComboBox<TreeSelectDataItem> {
   }
 
   /**
-   * Filter the internal items by query
-   * Changes items' hidden state
-   *
+   * Filter the internal items by query changes items' hidden state
```
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree-select/index.ts]: Updated

File: packages/elements/src/tree/elements/tree.ts (hunk 1 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/elements/tree.ts]: Fixed

File: packages/elements/src/tree/helpers/filter.ts (hunk 5 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/helpers/filter.ts]: Fixed

File: packages/elements/src/tree/__demo__/index.html (hunk 7 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/__demo__/index.html]: Fixed

File: packages/elements/src/tree/elements/tree.ts [hunk 13]
```diff
@@ -254,10 +268,136 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
    * @override
    */
   protected update (changedProperties: PropertyValues): void {
+    super.update(changedProperties);
```
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/elements/tree.ts:271]: In my view, in most cases `super` should be called before performing tasks because the `super class` might have some performing tasks e.g. initial, prepare something first so if we call `super` after, the `super class` a component might be a break in some cases.

File: packages/elements/src/tree/elements/tree.ts (hunk 3 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/elements/tree.ts]: I just thought of there are two cases to use this utility method `addNestedItemsToRender` 
1. Add all nested items and include all hidden items
2. Add all nested items but not include all hidden items

I updated the method to support two cases by add param `includeHidden`

**Comment by Sakchai-Refinitiv** [comment]
> Hey @Sakchai-Refinitiv,
> 
> Please address these highlighted issues.
> 
> Can you also check the linting issues. I know they are not part of your changes, however, they can easily be fixed.

I can't see any listing issue for now.

File: packages/elements/src/tree/elements/tree.ts (hunk 3 above)
Comment (TremayneChrist) [reply packages/elements/src/tree/elements/tree.ts]: If a user sets the item to be hidden in the original data object, it should never show in the results.

File: packages/elements/src/tree/elements/tree.ts (hunk 4 above)
Comment (TremayneChrist) [reply packages/elements/src/tree/elements/tree.ts]: Why is that?

File: packages/elements/src/tree/helpers/types.ts (hunk 6 above)
Comment (TremayneChrist) [reply packages/elements/src/tree/helpers/types.ts:3]: Is this an internal type only?

Comment (TremayneChrist) [reply packages/elements/src/tree/helpers/types.ts:3]: Or will users use this type?

File: packages/elements/src/tree-select/__test__/tree-select.filter.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (TremayneChrist) [reply packages/elements/src/tree-select/__test__/tree-select.filter.test.js:63]: It doesn't matter, the test should read well.

File: packages/elements/src/tree/helpers/types.ts (hunk 6 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/helpers/types.ts:3]: Oh, I got it, this is the same default parameter concept. yes, it shouldn't be removed I'll fix it.

File: packages/elements/src/tree/elements/tree.ts (hunk 4 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/elements/tree.ts]: OK, checking the hidden is not need anymore as you said because of the fixing code in the above conversation. I'll remove it.

Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/elements/tree.ts]: Removed

File: packages/elements/src/tree-select/__test__/tree-select.filter.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree-select/__test__/tree-select.filter.test.js:63]: I think the `openUpdated` is clear for me.

File: packages/elements/src/tree-select/__test__/tree-select.filter.test.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree-select/__test__/tree-select.filter.test.js]: OK, I get what you mean. fixed

File: packages/elements/src/tree-select/__test__/tree-select.filter.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree-select/__test__/tree-select.filter.test.js]: Fixed

File: packages/elements/src/tree/helpers/types.ts [hunk 14]
```diff
@@ -17,3 +17,13 @@ export interface TreeDataItem extends DataItem {
...
+
+
+/**
+ * Predicate callback
+ * Matches item against filter function
+ *
+ * @param item Item to filter
+ * @return Does item match filter
+ */
+export type TreeFilter<T extends TreeDataItem> = (item: T) => boolean;
```
Comment (TremayneChrist) [inline packages/elements/src/tree/helpers/types.ts:29]: Should you set the default type here, too?

File: packages/elements/src/tree/__demo__/index.html (hunk 7 above)
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/__demo__/index.html:167]: Fixed

**Comment by Sakchai-Refinitiv** [comment]
@goremikins If you don't have any concern, please merge this PR I already fixed all code of Trem comments 😀

File: packages/elements/src/tree/elements/tree.ts [hunk 15]
```diff
@@ -254,10 +268,138 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
...
+    }
+  }
+
+  /**
+   * Utility method
+   * Adds descendants for each item passed
+   * @param items List of child items
+   * @returns {void}
+   */
+  public addItemDescendantsToRender (items: T[]): void {
```
Comment (goremikins) [inline packages/elements/src/tree/elements/tree.ts:329]: Should this method be `protected`?

File: packages/elements/src/tree/helpers/filter.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+      query = el.query || '';
+      queryRegExp = new RegExp(query.replace(/(\W)/g, '\\$1'), 'i');
+    }
+    return queryRegExp;
+  };
+
+  // return scoped custom filter
+  return (item): boolean => {
+    const regex = getRegularExpressionOfQuery();
+    const result = regex.test(String((item as unknown as TreeDataItem).label));
```
Comment (goremikins) [inline packages/elements/src/tree/helpers/filter.ts:29]: String(null) === 'null' and String(undefined) === 'undefined'. It does not make sense to filter on null or undefined. I label is not present, it should tested agains empty string ('') or in other words simply excluded from the result.

File: packages/elements/src/tree/elements/tree.ts [hunk 17]
```diff
@@ -254,10 +268,138 @@ export class Tree<T extends TreeDataItem = TreeDataItem> extends List<T> {
...
+      }
+    });
+  }
+
+  /**
+   * Utility method
+   * Add nested children of item list
+   * @param items List of items
+   * @param excludeItems List of exclude items
+   * @param includeHidden Include hidden items
```
Comment (goremikins) [inline packages/elements/src/tree/elements/tree.ts:355]: `@param [includeHidden=false] Include hidden items`

File: packages/elements/src/tree/helpers/filter.ts [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+  // return scoped custom filter
+  return (item): boolean => {
+    const label = (item as unknown as TreeDataItem)?.label;
+    if (!label) {
+      return false;
+    }
+
+    const regex = getRegularExpressionOfQuery();
+    const result = regex.test(label);
+    // this example uses global scope, so the index needs resetting
```
[the remark this answers is not in the document]
Comment (Sakchai-Refinitiv) [reply packages/elements/src/tree/helpers/filter.ts]: Fixed

**Review by goremikins — APPROVED** [review]
